src/dragonball/src/dbs_pci/src/bus.rs

```diff
@@ -0,0 +1,415 @@
+        }
+    }
+
+    fn assign_default_device_id(devices: &mut IntervalTree<Arc<dyn PciDevice>>) {
+        // At present, device id means slot id of device.
+        // TODO: Code logic needs to be optimized to support multifunction devices.
+        devices.insert(Range::new(0x0u8, 0x1fu8), None);
+    }
+
+    /// Get bus ID for this PCI bus instance.
+    pub fn bus_id(&self) -> u8 {
+        self.bus_id
+    }
+
+    /// Allocate an unused PCI device ID.
+    ///
+    /// # Arguments:
+    /// * - `device_id`: allocate the specified device ID if it's valid.
+    pub fn allocate_device_id(&self, device_id: Option<u8>) -> Option<u8> {
+        // A PCI device supports 8 functions at most, aligning on 8 means only using function 0.
+        let mut constraint = Constraint::new(1u64).align(1u64);
+        if let Some(id) = device_id {
+            constraint = constraint.min(id as u64).max(id as u64);
+        }
+
+        debug!(
+            "allocate device id constraint size: {}, min: {}, max: {}",
+            constraint.size, constraint.min, constraint.max
+        );
+        // Do not expect poisoned lock here.
+        self.devices
+            .write()
+            .expect("poisoned RwLock() for PCI bus")
+            .allocate(&constraint)
+            .map(|e| e.min as u8)
+    }
+
+    /// Free the previously allocated device id and return data associated with the id.
+    pub fn free_device_id(&self, device_id: u32) -> Option<Arc<dyn PciDevice>> {
+        if device_id > 0x1f {
+            return None;
+        }
+        // Safe to unwrap because no legal way to generate a poisoned RwLock.
+        self.devices
+            .write()
+            .unwrap()
+            .free(&Range::new(device_id as u64, device_id as u64))
+    }
+
+    /// Add a child PCI device to the bus.
+    pub fn register_device(&self, device: Arc<dyn PciDevice>) -> Result<()> {
+        // Do not expect poisoned lock here.
+        let device_id = device.id();
+        let mut devices = self.devices.write().expect("poisoned lock for PCI bus");
+
+        debug!("add device id {} to bus", device_id);
+        let old = devices.update(&Range::new(device_id, device_id), device.clone());
+        assert!(old.is_none());
+
+        Ok(())
+    }
+
+    /// Get the device instance associated with the `device_id`.
+    pub fn get_device(&self, device_id: u32) -> Option<Arc<dyn PciDevice>> {
+        if device_id > 0x1f {
+            return None;
+        }
+        let devices = self.devices.read().unwrap();
+        match devices.get(&Range::new(device_id as u64, device_id as u64)) {
+            Some(NodeState::Valued(d)) => Some(d.clone()),
+            _ => None,
+        }
+    }
+
+    /// Read from PCI device configuration space.
+    pub fn read_config(&self, dev: u32, func: u32, offset: u32, data: &mut [u8]) {
+        if dev > 0x1f
+            || func != 0
+            || offset >= 0x1000
+            || offset & (data.len() - 1) as u32 & 0x3 != 0
```

Comment:
How about encapsulating this to an inline function? Since it is used twice.

Reply:
this is a really nice suggestion to reduce code duplication. thanks!